### PR TITLE
env(zjow): add env gym_pybullet_drones

### PR DESCRIPTION
**Pull request by zjowowen**
## Description
Gym_pybullet_drones environment is added into DI-zoo.
Two simple built-in single-agent tasks are included, takeoff-aviary-v0 and flythrugate-aviary-v0.
Multi-agent tasks are temporarily not supported.

## Related Issue
A special version of Gym_pybullet_drones repository for fixing video generation problems and gym>=0.25.0 is provided here:
https://github.com/zjowowen/gym-pybullet-drones/tree/gym-update

### CODE REVIEW

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from easydict import EasyDict
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py:1]: 参考我们之前的[config规范](https://di-engine-docs.readthedocs.io/zh_CN/latest/20_spec/config_spec_zh.html#id4), 文件重命名为`*_onppo_*.py`? 相关变量名也改为`*onppo*`？更加清晰一些

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            adv_norm=True,
+            value_norm=True,
+        ),
+        collect=dict(
+            n_sample=2048,
+            unroll_len=1,
+            discount_factor=0.99,
+            gae_lambda=0.97,
+        ),
+        #eval=dict(evaluator=dict(eval_freq=5000, )),
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py:42]: why eval_freq=5? delete unused comments.

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+flythrugate_ppo_config = dict(
+    exp_name='flythrugate_ppo_seed0',
+    env=dict(
+        manager=dict(shared_memory=False, reset_inplace=True),
+        env_id='flythrugate-aviary-v0',
+        norm_obs=dict(use_norm=False, ),
+        norm_reward=dict(use_norm=False, ),
+        collector_env_num=1,
+        evaluator_env_num=1,
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py:11]: 为什么是1，不设置`*_env_num=8,`呢？

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+)
+takeoffaviary_ppo_config = EasyDict(takeoffaviary_ppo_config)
+main_config = takeoffaviary_ppo_config
+
+takeoffaviary_ppo_create_config = dict(
+    env=dict(
+        type='gym_pybullet_drones',
+        import_names=['dizoo.gym_pybullet_drones.envs.gym_pybullet_drones_env'],
+    ),
+    env_manager=dict(type='base'),
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py:59]: why base?

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        use_act_scale=True,
+        n_evaluator_episode=2,
+        stop_value=0,
+        action_type="VEL",
+    ),
+    policy=dict(
+        cuda=True,
+        recompute_adv=True,
+        #load_path="./takeoffaviary_ppo_seed0/ckpt/ckpt_best.pth.tar",
+        #load_path="./ckpt_best-reward-1st-20.pth.tar",
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py:21]: 删除多余的注释

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        load_path="./ckpt_best-vel-1st-23-22.pth.tar",
+        model=dict(
+            obs_shape=12,
+            action_shape=4,
+            action_space='continuous',
+        ),
+        action_space='continuous',
+        learn=dict(
+            #是否share encoder
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py:30]: 写成类似这样？`TODO(zjow): whether to share the encoder?`

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            #liujie
+            epoch_per_collect=10,  #reduce
+            batch_size=64,
+            learning_rate=3e-4,  #tune; pytorch lr scheduler
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py:34]: #后有空格，yapf format

File: dizoo/gym_pybullet_drones/entry/flythrugate_ppo_eval.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    cfg.env['record'] = True
+    cfg.env['gui'] = True
+    cfg.env['print_debug_info'] = True
+    cfg.env['plot_observation'] = True
+
+    evaluator_env = BaseEnvManager(
+        env_fn=[lambda: GymPybulletDronesEnv(cfg.env) for _ in range(evaluator_env_num)], cfg=cfg.env.manager
+    )
+    #evaluator_env.enable_save_replay(cfg.env.replay_path)
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/entry/flythrugate_ppo_eval.py:41]: 注意空格的格式，增加注释

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+                'physics': Physics.PYB,
+                'freq': 240,
+                'aggregate_phy_steps': 1,
+                'gui': False,
+                'record': False,
+                'obs': ObservationType.KIN,
+                'act': ActionType.RPM
+            }
+        else:
+            #TODO for multi drones envs
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:105]: `# TODO(zjow): develop envs that support multi drones.`

File: dizoo/gym_pybullet_drones/envs/test_ding_env.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    print(env.action_space)
+    print(env.observation_space)
+    print(env.reward_space)
+    print("[INFO] Action space:", env.action_space)
+    print("[INFO] Observation space:", env.observation_space)
+
+    done = False
+    while not done:
+        action = env.action_space.sample()
+        obs, reward, done, info = env.step(action)
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/envs/test_ding_env.py:30]: 增加一些关于obs, reward shaoe的assert

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    elif action_str == "ONE_D_PID":
+        return ActionType.ONE_D_PID
+    elif action_str == "ONE_D_RPM":
+        return ActionType.ONE_D_RPM
+    else:
+        raise ValueError('Invalid action type.')
+
+
+@ENV_REGISTRY.register('gym_pybullet_drones')
+class GymPybulletDronesEnv(BaseEnv):
```
Comment (puyuan1996) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:87]: 增加overview注释，简介环境的定义，功能，用途等？

File: dizoo/gym_pybullet_drones/envs/test_ori_env.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    env = gym.make("takeoff-aviary-v0")
+    print(env.action_space)
+    print(env.observation_space)
+    print("[INFO] Action space:", env.action_space)
+    print("[INFO] Observation space:", env.observation_space)
+    env.reset()
+    done = False
+    while not done:
+        action = env.action_space.sample()
+        obs, reward, done, info = env.step(action)
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/test_ori_env.py:19]: 下面这个gif只截取其中核心画面部分吧，可以用ffmpeg来搞

File: dizoo/gym_pybullet_drones/envs/test_ding_env.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    cfg = EasyDict(cfg)
+    env = GymPybulletDronesEnv(cfg)
+
+    print(env.action_space)
+    print(env.observation_space)
+    print(env.reward_space)
+
+    env.reset()
+
+    print(env.action_space)
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/test_ding_env.py:21]: 测试一定要有assert，而且print能用assert代替的一定要用assert，这样不跑这个测试也能知道具体测试的效果

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        return BaseEnvTimestep(obs, rew, done, info)
+
+    def enable_save_replay(self, replay_path: Optional[str] = None) -> None:
+        if replay_path is None:
+            replay_path = './video'
+        self._replay_path = replay_path
+
+    def random_action(self) -> np.ndarray:
+        random_action = self.action_space.sample().astype(np.float32)
+        return random_action
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:246]: 直接return即可吧，没有必要多一个中间变量

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+            self._env.close()
+        self._init_flag = False
+
+    def seed(self, seed: int, dynamic_seed: bool = True) -> None:
+        self._seed = seed
+        self._dynamic_seed = dynamic_seed
+        np.random.seed(self._seed)
+
+    def step(self, action: np.ndarray) -> BaseEnvTimestep:
+        action = action.astype('float32')
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:213]: 为什么这里要变成float32，浮点数不本来就是float32吗

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+                'freq': 240,
+                'aggregate_phy_steps': 1,
+                'gui': False,
+                'record': False,
+                'obs': ObservationType.KIN,
+                'act': ActionType.RPM
+            }
+        else:
+            #TODO for multi drones envs
+            self.env_kwargs = {
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:106]: 写成default_config那种模式吧

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    },
+    "flythrugate-aviary-v0": {
+        "observation_space": gym_pybullet_drones_observation_space(12, minimum=-1, maximum=1),
+        "action_space": gym_pybullet_drones_action_space(drone_num=1, minimum=-1, maximum=1),
+        "reward_space": gym_pybullet_drones_reward_space()
+    },
+}
+
+
+def action_type(action_str):
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:65]: 可以尝试用枚举类来实现？

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+    if type_of_action in [ActionType.RPM, ActionType.DYN, ActionType.VEL]:
+        return 4
+    elif type_of_action == ActionType.PID:
+        return 3
+    elif type_of_action == ActionType.TUN:
+        return 6
+    elif type_of_action in [ActionType.ONE_D_DYN, ActionType.ONE_D_PID, ActionType.ONE_D_RPM]:
+        return 1
+    else:
+        raise ValueError('Invalid action type.')
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:31]: 要把invalid的type也放在错误信息里呀，否则不知道错的是啥

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+            value_weight=0.5,
+            entropy_weight=0.0,  #0.001
+            clip_ratio=0.2,  #0.1
+            adv_norm=True,
+            value_norm=True,
+        ),
+        collect=dict(
+            n_sample=2048,
+            unroll_len=1,
```
Comment (PaParaZz1) [inline dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py:43]: 简化不常修改的字段

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py (hunk 1 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py (hunk 2 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py (hunk 3 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/flythrugate_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py (hunk 4 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py (hunk 5 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py (hunk 7 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/entry/flythrugate_ppo_eval.py (hunk 8 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/entry/flythrugate_ppo_eval.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py (hunk 9 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/test_ding_env.py (hunk 10 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/test_ding_env.py:19]: fixed.

File: dizoo/gym_pybullet_drones/envs/test_ori_env.py (hunk 12 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/test_ori_env.py:15]: fixed.

File: dizoo/gym_pybullet_drones/envs/test_ding_env.py (hunk 13 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/test_ding_env.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py (hunk 14 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py (hunk 18 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:34]: fixed.

File: dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py (hunk 19 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/config/takeoffaviary_ppo_config.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py (hunk 15 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py]: fixed.

File: dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py (hunk 11 above)
Comment (zjowowen) [reply dizoo/gym_pybullet_drones/envs/gym_pybullet_drones_env.py:80]: fixed.

**Review by PaParaZz1 — APPROVED** [review]
